### PR TITLE
Support multiple accounts for remote urls that differ in path

### PR DESCRIPTION
**Pull request by niik**
ref https://github.com/desktop/desktop/issues/18586
based on https://github.com/desktop/desktop/pull/18627

## Description

Changes our account lookup logic to use the entire remote url (if we get it by Git, i.e. the `useHttpPath` config option must be turned on for the host).

Prior to 3.3.15 a [long lived regex mistake](https://github.com/desktop/desktop/blob/38a17e98b2ab66dd4c60ce9d3550968ddd7a9948/app/src/lib/remote-parsing.ts#L29-L32) in the `parseRemote` function caused the hostname for Azure Devops urls to include the first two segments of the path. When we called [`getGenericHostname`](https://github.com/desktop/desktop/blob/1cdc6b9477bf1c07c6a41d8e9b644e531b9f2d40/app/src/lib/generic-git-auth.ts#L7-L19) on a remote url like `https://dev.azure.com/org/repo/_git/repo` the hostname would get parsed as `dev.azure.com/org/repo` thus creating a unique key for storing a token for a specific repository.

When we resolved that flaw all we had access to was the hostname which broke the credential storage for users having multiple repositories across multiple orgs on Azure Devops.

This change will have Desktop honor the full remote url when Git is configured to provide it. We intend to make a change to https://github.com/desktop/dugite to set `useHttpPath` for dev.azure.com just like [Git for Windows does](https://github.com/git-for-windows/git/blob/79cd17fc78816ac676b4bdb3ea3e98e53413179e/scalar.c#L145)

In this work we've also had to contend with URLs containing usernames such as `https://USERNAME@dev.azure.com/username/repo/_git/repo`. For these remote urls we've disabled the username input in our password prompt and respect the username from the url.

Screenshot of a password prompt dialog in GitHub Desktop

## Release notes

Notes:

### CODE REVIEW

**Comment by tidy-dev** [comment]
I was thinking we had a scenario like `https://USERNAME@dev.azure.com/username/repo/_git/repo` where the username was "xxxyyy" and in the url it was "xxx.yyy". Is the one in the username always correct and the provided one incorrect?

**Comment by niik** [comment]
The short answer is yes, the provided username before the @ is correct since that's what Git will use. That doesn't mean the other won't work, that's an implementation detail we know nothing about but we're aligning with Git's behavior here

File: app/src/models/git-account.ts [hunk 1]
```diff
@@ -7,4 +7,7 @@ export interface IGitAccount {
 
   /** The endpoint with which the user is authenticating. */
   readonly endpoint: string
+
+  /** The  */
```
Comment (tidy-dev) [inline app/src/models/git-account.ts]: :P `The` sounds kind of ominous all by itself.

**Review by tidy-dev — DISMISSED** [review]
💖  Thanks for all the work on this issue.

**Review by tidy-dev — DISMISSED** [review]

**Review by tidy-dev — DISMISSED** [review]

**Review by tidy-dev — DISMISSED** [review]

**Comment by danielkoek** [comment]
Ref #18667, this will be the fix, might be handy to update the docs on https://github.com/desktop/desktop/blob/development/docs/integrations/azure-devops.md to clarify that it is org, not username that is being used as the first path segment. (which in the case of a personal project is the username)


It would be also possible to check the first part of the URL https://<org>@dev.azure.com instead of the first segment of the path. in case this still causes regression bugs

**Review by sergiou87 — DISMISSED** [review]

**Review by sergiou87 — APPROVED** [review]

**Review by sergiou87 — APPROVED** [review]

**Pull request merged**